File: common-golang/common-java/src/main/java/com/common/kafka/KafkaClient.java
package com.common.kafka;

import com.common.errors.GeneralError;
import com.common.message.Message;
import com.common.message.Response;
import com.common.utils.JsonUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.kafka.core.KafkaTemplate;

import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

public class KafkaClient {
    private static final Logger log = LoggerFactory.getLogger(KafkaClient.class);

    private final KafkaConfig config;
    private final KafkaTemplate<String, String> kafkaTemplate;
    private final ConcurrentHashMap<String, CompletableFuture<Message>> pendingRequests;

    public KafkaClient(KafkaConfig config, KafkaTemplate<String, String> kafkaTemplate) {
        this.config = config;
        this.kafkaTemplate = kafkaTemplate;
        this.pendingRequests = new ConcurrentHashMap<>();
    }

    public Response sendRequest(String topic, String uri, Object data) throws Exception {
        String txnId = MessageContextHolder.getTransactionId() != null ? 
                       MessageContextHolder.getTransactionId() : UUID.randomUUID().toString();

        Message msg = Message.newRequest(uri, config.getClusterId(), txnId, config.responseTopic(), data);

        CompletableFuture<Message> future = new CompletableFuture<>();
        pendingRequests.put(msg.getMessageId(), future);

        try {
            sendMessageInternal(topic, msg);
            Message respMsg = future.get(config.getRequestTimeoutMs(), TimeUnit.MILLISECONDS);
            return unwrapResponse(respMsg);
        } catch (TimeoutException e) {
            throw GeneralError.newTimeout(topic + "/" + uri);
        } finally {
            pendingRequests.remove(msg.getMessageId());
        }
    }

    public void sendMessage(String topic, String uri, Object data) throws Exception {
        String txnId = MessageContextHolder.getTransactionId() != null ? 
                       MessageContextHolder.getTransactionId() : UUID.randomUUID().toString();
        Message msg = Message.newRequest(uri, config.getClusterId(), txnId, null, data);
        msg.setMessageType(com.common.message.MessageType.MESSAGE);
        sendMessageInternal(topic, msg);
    }

    public void sendResponse(Message original, Object data) throws Exception {
        if (original.getResponseDestination() == null || original.getResponseDestination().getTopic() == null) {
            return;
        }
        Message resp = Message.newResponse(original, config.getClusterId(), data);
        sendMessageInternal(original.getResponseDestination().getTopic(), resp);
    }

    private void sendMessageInternal(String topic, Message msg) throws Exception {
        String json = JsonUtils.toJson(msg);
        log.debug("kafka send topic: {} uri: {} msgId: {}", topic, msg.getUri(), msg.getMessageId());
        kafkaTemplate.send(topic, msg.getMessageId(), json).get();
    }

    public void handleResponse(Message msg) {
        CompletableFuture<Message> future = pendingRequests.get(msg.getMessageId());
        if (future != null) {
            future.complete(msg);
        } else {
            log.warn("response for unknown request msgId: {}", msg.getMessageId());
        }
    }

    private Response unwrapResponse(Message msg) throws Exception {
        Response resp = JsonUtils.parseData(msg.getData(), Response.class);
        if (resp != null && resp.hasError()) {
            throw new GeneralError(resp.getStatus().getCode())
                    .withMessageParams(resp.getStatus().getMessageParams())
                    .withSource(resp.getStatus().getSource())
                    .withParams(resp.getStatus().getParams());
        }
        return resp;
    }
}

File: common-js/KafkaClient.js
const { Kafka } = require('kafkajs');
const { randomUUID } = require('crypto');
const { Message, Response } = require('./Message');
const MessageContextHolder = require('./MessageContextHolder');

class GeneralError extends Error {
    constructor(code, msgParams = null) {
        super(code);
        this.code = code;
        this.messageParams = msgParams;
        this.source = null;
        this.params = null;
    }

    static newTimeout(resource) {
        return new GeneralError('TIMEOUT', [resource]);
    }
}

class KafkaClient {
    constructor(config) {
        this.config = config;
        this.kafka = new Kafka({
            clientId: config.nodeId,
            brokers: config.brokers
        });
        this.producer = this.kafka.producer();
        this.pendingRequests = new Map();
        
        this._isConnected = false;
    }

    async connect() {
        if (!this._isConnected) {
            await this.producer.connect();
            this._isConnected = true;
        }
    }

    async sendRequest(topic, uri, data, headers = {}) {
        await this.connect();

        const ctxTxnId = MessageContextHolder.getTransactionId();
        const txnId = ctxTxnId ? ctxTxnId : randomUUID();

        const msg = Message.newRequest(uri, this.config.clusterId, txnId, this.config.responseTopic(), data);
        msg.headers = headers;
        
        return new Promise(async (resolve, reject) => {
            const timeoutId = setTimeout(() => {
                this.pendingRequests.delete(msg.messageId);
                reject(GeneralError.newTimeout(`${topic}/${uri}`));
            }, this.config.requestTimeoutMs);

            this.pendingRequests.set(msg.messageId, { resolve, reject, timeoutId });

            try {
                await this._sendMessageInternal(topic, msg);
            } catch (err) {
                clearTimeout(timeoutId);
                this.pendingRequests.delete(msg.messageId);
                reject(err);
            }
        });
    }

    async sendMessage(topic, uri, data) {
        await this.connect();
        const ctxTxnId = MessageContextHolder.getTransactionId();
        const txnId = ctxTxnId ? ctxTxnId : randomUUID();
        
        const msg = Message.newRequest(uri, this.config.clusterId, txnId, null, data);
        msg.messageType = 'MESSAGE';
        await this._sendMessageInternal(topic, msg);
    }

    async sendResponse(original, data) {
        await this.connect();
        if (!original.responseDestination || !original.responseDestination.topic) {
            return;
        }
        const resp = Message.newResponse(original, this.config.clusterId, data);
        await this._sendMessageInternal(original.responseDestination.topic, resp);
    }

    async _sendMessageInternal(topic, msg) {
        const json = JSON.stringify(msg);
        await this.producer.send({
            topic: topic,
            messages: [{ key: msg.messageId, value: json }]
        });
    }

    handleResponse(msg) {
        const mapped = this.pendingRequests.get(msg.messageId);
        if (mapped) {
            this.pendingRequests.delete(msg.messageId);
            clearTimeout(mapped.timeoutId);
            try {
                const resp = this._unwrapResponse(msg);
                mapped.resolve(resp);
            } catch(e) {
                mapped.reject(e);
            }
        } else {
            console.warn(`response for unknown request msgId: ${msg.messageId}`);
        }
    }

    _unwrapResponse(msg) {
        // Cast generic object to Response
        const resp = new Response();
        Object.assign(resp, msg.data);
        
        if (resp.hasError()) {
            const err = new GeneralError(resp.status.code, resp.status.messageParams);
            err.source = resp.status.source;
            err.params = resp.status.params;
            throw err;
        }
        return resp;
    }
}

module.exports = { KafkaClient, GeneralError };
